backends/imgui_impl_dx9.cpp

```diff
@@ -70,6 +75,435 @@ static ImGui_ImplDX9_Data* ImGui_ImplDX9_GetBackendData()
     return ImGui::GetCurrentContext() ? (ImGui_ImplDX9_Data*)ImGui::GetIO().BackendRendererUserData : NULL;
 }
 
+// Shader pipeline
+static bool ImGui_ImplDX9_Shader_CreatePipeline()
+{
+    ImGui_ImplDX9_Data* bd = ImGui_ImplDX9_GetBackendData();
+    
+    D3DCAPS9 caps; ZeroMemory(&caps, sizeof(D3DCAPS9));
+    if (D3D_OK != bd->pd3dDevice->GetDeviceCaps(&caps))
+        return false;
+    if (caps.VertexShaderVersion < D3DVS_VERSION(2,0) || caps.PixelShaderVersion < D3DPS_VERSION(2, 0))
+        return false;
+    
+    // Input layout of default ImDrawVert
+    static const D3DVERTEXELEMENT9 InputLayout[4] = {
+        { 0,  0, D3DDECLTYPE_FLOAT2  , D3DDECLMETHOD_DEFAULT, D3DDECLUSAGE_POSITION, 0 },
+        { 0,  8, D3DDECLTYPE_FLOAT2  , D3DDECLMETHOD_DEFAULT, D3DDECLUSAGE_TEXCOORD, 0 },
+        { 0, 16, D3DDECLTYPE_D3DCOLOR, D3DDECLMETHOD_DEFAULT, D3DDECLUSAGE_COLOR   , 0 },
+        D3DDECL_END(),
+    };
+    if (D3D_OK != bd->pd3dDevice->CreateVertexDeclaration(InputLayout, &bd->pInputLayout))
+        return false;
+    
+    // Shaders
+    #define CODE(b1, b2, b3, b4) 0x##b1, 0x##b2, 0x##b3, 0x##b4
+    static const BYTE VertexShaderByteCode[] = {
+        // HLSL source code
+        /*
+            float4x4 mvp : register(c0);
+            
+            struct VS_Input
+            {
+                float2 pos : POSITION0;
+                float2 uv  : TEXCOORD0;
+                float4 col : COLOR0;
+            };
+            
+            struct VS_Output
+            {
+                float4 pos : POSITION;
+                float2 uv  : TEXCOORD0;
+                float4 col : COLOR0;
+            };
+            
+            VS_Output main(VS_Input input)
+            {
+                VS_Output output;
+                
+                output.pos = mul(mvp, float4(input.pos.x, input.pos.y, 0.0f, 1.0f));
+                output.uv = input.uv;
+                output.col = input.col;
+                
+                return output;
+            };
+        */
+        // Precompile with Visual Studio 2019 HLSL compiler, vs_2_0
+        CODE(00,02,FE,FF),
+        CODE(FE,FF,1E,00),
+        CODE(43,54,41,42),
+        CODE(1C,00,00,00),
+        CODE(4B,00,00,00),
+        CODE(00,02,FE,FF),
```

Comment:
Suggest removing the macros and using raw 0xXX data, with 16-bytes per lines. This will reduce the line count by 4 and remove a subtle layer of complexity.
```cpp
CODE(00,02,FE,FF),
CODE(FE,FF,1E,00),
CODE(43,54,41,42),
CODE(1C,00,00,00),
```
into
```
0x00, 0x02, 0xFE, 0xFF, 0xFE, 0xFF, 0x1E, 0x00, 0x43, 0x54, 0x41, 0x42, 0x1C, 0x00, 0x00, 0x00,
```

Reply:
Suggest adding a comment with an explicit command-line to turn the shader source into binary data.
I realize that MAYBE the output of that compiler is to output `CODE(XX,XX,XX,XX)` but as it is unlikely to be rebuild (our other shaders have stayed stables for many years) this is mostly as reference.

Reply:
I finally decided to put them on the same line, and do not use hex, this helps reduce the number of source code characters